### PR TITLE
Ability to run peer pod enablement on an existing cluster

### LINKED ISSUE #21: doc | ibmcloud: Inconsistency in default values for `cos_service_instance_name `
From https://github.com/confidential-containers/cloud-api-adaptor/pull/18#discussion_r875737240:
> In the next section the doc says:
> > cos_service_instance_name is the COS Service Instance Name, optional, default to value based on your cluster_name > tfvar if not set, ${cluster_name}-cos-service-instance.
>
> I think we should make the defaults of these line up, otherwise it will get confusing.

We now have inconsistency in the default values of the `cos_service_instance_name` in the two sections it is used:
- In creating the COS instance we have:
  > cos_service_instance_name in variables.tf has the default value 
- In building the pod VM image and uploading it we have:
  > cos_service_instance_name is the COS Service Instance Name, optional, default to value based on your cluster_name tfvar if not set, ${cluster_name}-cos-service-instance.
  
This is not a good user experience, so we have a few choices:
- Either get rid of the defaults and stop them being optional altogether
- Change both defaults to `cos-image-instance`. I don't think there is a requirement for the cos service instance name to be unique across users, so this should work
- Change both defaults to `<cluster_name>-cos-image-instance`. Unless there is a way to read the `cluster_name` value from the cluster terraform.tfvars, I don't think this is a good solution, as there isn't another reason for cluster name to be required in the cos instance creation, so it's letting a user not supply one value by making them supply another
- Any other solution someone smarter than me can think of?

### LINKED ISSUE #22: As a peer pods tester I want the ability to run peer pod enablement on an existing cluster
## Description
We want to split the Ansible playbook in `terraform/cluster/ansible` into two new Ansible playbooks so we have the benefit running one for the initial cluster setup, and another for Peer Pod enablement. 

## Use case
_As a peer pods tester I want the ability to run peer pod enablement on an existing cluster. This will be the setup for an operator that runs on a Kube cluster._

## Task List
- [x] In the `terraform/cluster/ansible` playbook, split the initial cluster setup code from the Peer Pod enablement code into two playbooks.
- [x] Determine how the two new playbooks are going to run, as to not disrupt the current demo instructions.
- [x] Change the Terraform template to run the two playbooks in the correct order.
- [x] Determine how the README in `terraform/`should be updated and update the README.
- [x] Raise a PR with the changes. The PR must be approved and merged before this story is closed.

## Performance Considerations
Increasing the number of manual steps goes against the idea of automation, when we are splitting a playbook into two to have control over which functionality we want, ensure that the steps to run the demo aren't made more complicated.

## Security Considerations 
API keys must not be shared or pushed to this repository.

### Questions
1. Are the two Ansible playbooks generated going to run back to back?

2. How would this issue be documented in the `/terraform/README`. We need to decide if the instructions to run the README to setup the demo will evolve into something new as the functionality of the Terraform scripts evolve.
 
3. How do we capture the benefits of one without increasing the number of steps for the demo? 

## Acceptance Criteria
**Given** I want to run Peer Pod enablement on an existing cluster
**When** I split the existing `/cluster/ansible/`  playbook up into separate functionality
**Then** the existing Terraform template can run the new Peer Pod enablement Ansible playbook against an existing cluster without running the initial cluster setup steps

**Scenario: Regression testing**
**Given** I want to create a new Kubernetes cluster and enable it with the peer-pod capability
**When** I follow the `Create a Kubernetes cluster` and peer pod enablement documentation 
**Then** I have a new, working Kubernetes cluster and it has the kata-shim, cloud-api-adaptor and all the required dependencies installed on it

**Given** I want to be able to create an x86 and s390x cluster
**When** the existing Ansible playbook is split into two and I run them
**Then** I maintain the option to choose an x86 or s390x cluster with peer-pod enablement for creation

### PR DESCRIPTION
**Pull request by Jonah-F**
automation: split cluster playbook into two
    
Completed work to split cluster/ansible playbook into
initial cluster setup and peer pod enablement.
I split dependencies up after flagging them as kube or
kata related, or both.
Runbooks execute consecutively so no changes to README.
    
Fixes #22
    
Signed-off-by: [Jonah-Farrow] <jonah.farrow@ibm.com>

### CODE REVIEW

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+          [Install]
+          WantedBy=multi-user.target
+
+    - name: Start fix-resolvconf service
+      systemd:
+        name: fix-resolvconf.service
+        state: started
+
+    - name: Install IBM Cloud CLI
```
Comment (mattarnoatibm) [inline ibmcloud/terraform/cluster/ansible/kube-playbook.yml:241]: We should test if the playbooks work when the Install IBM Cloud CLI and Install IBM Cloud CLI plugins tasks are only in the kata playbook.

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+          [Install]
+          WantedBy=multi-user.target
+
+    - name: Start containerd service
+      systemd:
+        name: containerd.service
+        state: started
+    
+    - name: Copy systemd unit file for fix-resolvconf
```
Comment (mattarnoatibm) [inline ibmcloud/terraform/cluster/ansible/kube-playbook.yml:219]: We should test if the playbooks work when the Copy systemd unit file for fix-resolvconf and Start fix-resolveconf service tasks are only in the kata playbook

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        - path: /etc/modules-load.d/k8s.conf
+          content: br_netfilter
+        - path: /etc/modules-load.d/peerpod.conf
+          content: vrf
+
+    - name: Load kernel modules
+      modprobe:
+        name: "{{ item }}"
+      with_items:
+        - br_netfilter
```
Comment (mattarnoatibm) [inline ibmcloud/terraform/cluster/ansible/kube-playbook.yml:100]: We should test the playbooks work if the Configure kernel modules to load at boot and load kernel modules task in the kubernetes playbook only sets up the br_netfilter kernel module.

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        - vrf
```
Comment (mattarnoatibm) [inline ibmcloud/terraform/cluster/ansible/kube-playbook.yml:101]: We should test the playbooks work if the Configure kernel modules to load at boot and load kernel modules task in the kata playbook only sets up the vrf kernel module.

**Review by mattarnoatibm — CHANGES_REQUESTED** [review]
For some of the duplicates, where the tasks are in both the Kubernetes and Kata playbooks we should test if the duplicates are needing in both playbooks. We established the containerd tasks do need to be duplicated. For the other duplicated tasks, which I've marked in comments, we should test if the duplication is needed or if the task can be in one playbook or the other only.

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml (hunk 3 above)
Comment (Jonah-F) [reply ibmcloud/terraform/cluster/ansible/kube-playbook.yml:98]: Tested and successfully reran Terraform template.
Remove vrf from Kube-playbook, and br_netfilter from Kata-playbook under Load kernel modules . Also, remove vrf configuration from Kube-playbook, and br_netfilter configuration from Kata-playbook under Configure kernel modules to load at boot.

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml (hunk 2 above)
Comment (Jonah-F) [reply ibmcloud/terraform/cluster/ansible/kube-playbook.yml]: Tested and successfully reran Terraform template.
Remove Copy systemd unit file for fix-resolvconf and Start fix-resolvconf service  (which uses systemd) from Kube-playbook

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml (hunk 1 above)
Comment (Jonah-F) [reply ibmcloud/terraform/cluster/ansible/kube-playbook.yml]: Tested and successfully reran Terraform template.
Remove IBMCloud CLI and plugin install from Kube-playbook

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml (hunk 4 above)
Comment (Jonah-F) [reply ibmcloud/terraform/cluster/ansible/kube-playbook.yml]: Removed vrf configuration from Kube-playbook, and br_netfilter configuration from Kata-playbook under Configure kernel modules to load at boot.

File: ibmcloud/terraform/cluster/ansible/kata-playbook.yml [hunk 5]
```diff
@@ -27,10 +18,6 @@
           - libdevmapper-dev
           - libseccomp-dev
           - pkg-config
-          - kubelet
-          - kubeadm
-          - kubectl
-          - kubernetes-cni
           - cri-tools
```
Comment (mattarnoatibm) [inline ibmcloud/terraform/cluster/ansible/kata-playbook.yml:21]: The cri-tools package is only in the https://apt.kubernetes.io kubernetes-xenial/main apt repository, so if the kata enablement playbook needs to install the cri tools it needs to keep the 'Add Kubernetes apt repository signing key' and 'Add Kubernetes apt repository' tasks.

If cri-tools is used to debug the kubelet, does installing the cri-tools package only belong in he Kubernetes enablement playbook?

**Review by mattarnoatibm — APPROVED** [review]
I've tested the two Ansible playbooks with the duplication removed. I've made one comment that if we want to keep installation of the cri-tools package in the kata playbook we need to keep adding the kubernetes apt repository in that playbook. The other comments I made about removing duplication from the playbook have been addressed.

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml (hunk 1 above)
Comment (mattarnoatibm) [reply ibmcloud/terraform/cluster/ansible/kube-playbook.yml]: I've tested the new playbooks and they work with installation of the IBM Cloud CLI only being in the kata playbook

File: ibmcloud/terraform/cluster/ansible/kube-playbook.yml (hunk 3 above)
Comment (mattarnoatibm) [reply ibmcloud/terraform/cluster/ansible/kube-playbook.yml:98]: I've tested the new playbooks and they work with enabling the br_netfilter kernel module only being in the kube playbook.

**Review by mattarnoatibm — CHANGES_REQUESTED** [review]
Sorry, meant to request a change.

File: ibmcloud/terraform/cluster/scripts/keygen.sh [hunk 6]
```diff
@@ -68,7 +68,7 @@ for host in "${hosts[@]}"; do
         trap stop 0
 
         ssh "${opts[@]}" -o ExitOnForwardFailure=yes -S "$ssh_ctl_sock" -L "$port:$host:22" -M -f -N -l root "$bastion"
-        ssh-copy-id "${opts[@]}" -i "$key.pub" -p "$port" root@localhost
+        ssh-copy-id -f "${opts[@]}" -i "$key.pub" -p "$port" root@localhost
```
Comment (mattarnoatibm) [inline ibmcloud/terraform/cluster/scripts/keygen.sh:71]: Changing the keygen.sh script is outside of the scope of the issue this PR is fixing, so we should be making this change under a separate PR and issue

Comment (Jonah-F) [reply ibmcloud/terraform/cluster/scripts/keygen.sh]: Thanks. I have reverted this change.

File: ibmcloud/terraform/cluster/ansible/kata-playbook.yml (hunk 5 above)
Comment (Jonah-F) [reply ibmcloud/terraform/cluster/ansible/kata-playbook.yml]: Tested and successfully reran Terraform template.
Removed the `cri-tools` dependency from the kata-playbook.
Test 1 result:
   Apply complete!

**Review by mattarnoatibm — APPROVED** [review]
Looks good to me.

**Review by stevenhorsman — APPROVED** [review]
I've run through creating a cluster using this code based on `x86` and it seemed to work without issues. Thanks Jonah!
